packages/cli/src/commands/build-cache/cache.ts

```diff
@@ -0,0 +1,172 @@
+/*
+ * Copyright 2020 Spotify AB
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import fs from 'fs-extra';
+import { resolve as resolvePath, relative as relativePath } from 'path';
+import { runPlain, runCheck } from '../../helpers/run';
+import { Options } from './options';
+import { extractArchive, createArchive } from './archive';
+
+const INFO_FILE = '.backstage-build-cache';
+
+export type CacheQueryResult = {
+  hit: boolean;
+  copy?: (outputDir: string) => Promise<void>;
+  archive?: (outputDir: string, maxEntries: number) => Promise<void>;
+};
+
+export type CacheKey = string[];
+
+type CacheEntry = {
+  key: CacheKey;
+  path: string;
+};
+
+type CacheInfo = {
+  key?: CacheKey;
+  entries?: CacheEntry[];
+};
+
+export class Cache {
+  static async read(options: Options) {
+    const repoPath = relativePath(options.repoRoot, process.cwd());
+    const location = resolvePath(options.cacheDir, repoPath);
+
+    // Make sure we don't have any uncommitted changes to the input, in that case we consider the cache to be missing
+    const noChanges = await runCheck(
+      `git diff --quiet HEAD -- ${options.inputs.join(' ')}`,
```

Comment:
i mean there's probably potential escaping issues here, like when there's spaces in a path or somebody puts the path `"$(rm -rf /)"` in there or so ... :) I think in general our runners shouldn't take a string argument but an array of strings instead that the shell deals with properly.

Buuuut good enough for now.

Reply:
No this was actually a mistake, each arg is supposed to be wrapped in `'`, and `'` are removed from the input.